### PR TITLE
Simple include/exclude filter for overmap search.

### LINKED ISSUE #23993: "River bank" name makes searching for "Bank" on map difficult
Game version: 0.C-28148-g3b40214

Operating system: Windows 7

Tiles or curses: tiles

Mods active: none

### Expected behavior
Searching for "bank" at the overmap should return a list of bank buildings in cities.

### Actual behavior
The search returns hundreds of useless "river banks", if you've discovered some river tiles previously.

### Steps to reproduce the behavior
Try searching the overmap for banks. Be sure to have some rivers discovered.

### Suggested solution
Either rename the "river bank" overmap tile to just "river", removing the conflict, or make the overmap search line take filtering operators, like the one in Shift+V does. The second variant would be much more useful, because 1. it would allow searching for multiple types of overmap tiles (eg "pharm, doctor, vet" at the start of Infected scenario), and 2. it would avoid similar problems in future and/or with localized versions.

### PR DESCRIPTION
**Pull request by OzoneH3**
Closes #23993

### CODE REVIEW

**Comment by ZhilkinSerg** [comment]
Let's add regex there!

**Comment by OzoneH3** [comment]
I'd like to but I'm kinda fearfull that it will introduce some kind of compile
problem on other systems, since it's not used anywhere as of now.

```wildcard_match()``` is another one I'd rather have written in regex
but oh well, maybe someday....

**Comment by ZhilkinSerg** [comment]
I'm concerned about performance:

I've just started a new game and was spawned at mine on Z level -5 with no surrounding tiles explored, entered `-unexplored` search term and the game hung searching for overmap terrains indefinitely long.

**Comment by OzoneH3** [comment]
Sounds like it's a already existing bug on lower z-levels where it tries to load new overmaps to find tiles to filter against.

**Comment by WhiteHusky** [comment]
Could always look out if a key, maybe 5, is being pressed and abort the search.
Or determine if the search is taking more than 5-10 seconds and query if it should continue or abort (and show the results it has so far)
Either way, might be a good way for someone to escape a hang for bugs like this. Might even mention in the log that it might be a potential bug and should be reported.

**Comment by OzoneH3** [comment]
Faster to just fix the bug than to implement something like that :P

**Comment by ZhilkinSerg** [comment]
> Sounds like it's a already existing bug on lower z-levels where it tries to load new overmaps to find tiles to filter against.

I am not sure about that - search does not hang on `master` for me. Will need to test this more when I have spare time.

**Comment by BrettDong** [comment]
However I am unable to reproduce the hang that you were experiencing. I started a _Bottom of a Mine_ scenario, got spawned on Z level -5, entered `-unexplored`, hit enter, and a box reading _No results found._ came out in a flash.

**Comment by ZhilkinSerg** [comment]
> However I am unable to reproduce the hang that you were experiencing. I started a Bottom of a Mine scenario, got spawned on Z level -5, entered -unexplored, hit enter, and a box reading No results found. came out in a flash.

Are you compiling on Linux, OS or Windows?

My VS builds with this PR are still hanging on search. :(

**Comment by BrettDong** [comment]
> Are you compiling on Linux, OS or Windows?

I'm compiling on macOS using Xcode. I think you can pause the game process in Visual Studio and see where it hangs.

**Comment by ZhilkinSerg** [comment]
So, it hangs on every search and when I hit break execution stops either on new `match_include_exclude` function:

image

or somewhere else (while checking overmap terrain name strings):

image

**Comment by ZhilkinSerg** [comment]
Tried MSYS build on Windows - there are errors during compilation.

```
src/cata_utility.cpp: In function 'bool match_include_exclude(const string&, std::__cxx11::string)':
src/cata_utility.cpp:57:33: error: comparison is always false due to limited range of data type [-Werror=type-limits]
         std::string term = iPos == std::string::npos ? filter : filter.substr( 0, iPos );
                            ~~~~~^~~~~~
src/cata_utility.cpp:71:18: error: comparison is always true due to limited range of data type [-Werror=type-limits]
         if( iPos != std::string::npos ) {
             ~~~~~^~~~~~
src/cata_utility.cpp:74:19: error: comparison is always true due to limited range of data type [-Werror=type-limits]
     } while( iPos != std::string::npos );
              ~~~~~^~~~~~
cc1plus.exe: all warnings being treated as errors
make: *** [Makefile:756: objwin/tiles/cata_utility.o] Error 1
```

File: src/cata_utility.cpp [hunk 1]
```diff
@@ -42,6 +42,40 @@ bool lcmatch( const std::string &str, const std::string &qry )
     return haystack.find( needle ) != std::string::npos;
 }
 
+bool match_include_exclude( const std::string &text, std::string filter )
+{
+    unsigned long iPos;
```
Comment (ZhilkinSerg) [inline src/cata_utility.cpp]: Changing this from `unsigned long` to `size_t` resolves hang in VS and MSYS compiles just fine.

**Pull request merged**